nengo_spa/compiler/grammar.py

```diff
@@ -0,0 +1,172 @@
+"""Definition of the Nengo SPA action rule grammar.
+
+All interaction with this module should use the `parse` function. The exact
+rules defined in this module may change.
+"""
+
+from io import BytesIO
+import sys
+
+from nengo_spa.compiler.parsing import (
+    AnyNumber, AtLeastOne, Group, Maybe, Peek, Rule, RuleMismatchError,
+    Terminal)
+import nengo_spa.compiler.tokens as tk
+from nengo_spa.internal.generators import Peekable
+
+
+def parse(code):
+    """Parses a block of code with the Nengo SPA action rule grammar.
+
+    Arguments
+    ---------
+    code : str
+        Code to parse.
+
+    Returns
+    -------
+    list
+        Represents the parse tree. Each non-terminal node is either represented
+        by a list of child nodes or a 2-tuple *(node_type, children)* with a
+        string *node_type* specifying the node type and a list of *children*.
+        The terminal nodes are `tokenize.TokenInfo` objects.
+    """
+    return actions.read(Peekable(tk.tokenize(
+        BytesIO(code.encode()).readline)))
+
+
+effect_line_delimiter = Terminal(type=tk.NEWLINE) | Terminal(type=tk.ENDMARKER)
+if sys.version_info[0] < 3:
+    class ConnectL2R(Rule):
+        def accept(self, tokens):
+            return [t[1] for t in tokens.peek(n=2)] == ['-', '>']
+
+        def read(self, tokens):
+            try:
+                return [next(tokens), next(tokens)]
+            except StopIteration:
+                raise RuleMismatchError(self, self, (tk.ENDMARKER, '', None))
+
+        def __str__(self):
+            return "'->'"
+
+        def __repr__(self):
+            return "ConnectL2R()"
+
+    connect_l2r = ConnectL2R()
+else:
+    connect_l2r = Terminal(type=tk.OP, string='->')
```

Comment:
Maybe add a comment here indicating that the "->" operator is present in later Python versions